src/utils/intersect.jl

```diff
@@ -0,0 +1,143 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    bentleyottmann(segments; digits)
+
+Compute pairwise intersections between n `segments`
+with `digits` precision in O(n⋅log(n)) time using
+Bentley-Ottmann sweep line algorithm.
+
+## References
+
+* Bentley & Ottmann 1979. [Algorithms for reporting and counting
+  geometric intersections](https://ieeexplore.ieee.org/document/1675432)
+"""
+function bentleyottmann(segments; kwargs...)
+  # orient segments
+  segs = map(segments) do s
+    a, b = extrema(s)
+    a > b ? (b, a) : (a, b)
+  end
+
+  # retrieve types
+  P = eltype(first(segs))
+  S = Tuple{P,P}
+
+  # initialization
+  𝒬 = BinaryTrees.AVLTree{P}()
+  ℛ = BinaryTrees.AVLTree{S}()
+  ℬ = Dict{P,Vector{S}}()
+  ℰ = Dict{P,Vector{S}}()
+  ℳ = Dict{P,Vector{S}}()
+  lookup = Dict{S,Int}()
+  for (i, (a, b)) in enumerate(segs)
+    BinaryTrees.insert!(𝒬, a)
+    BinaryTrees.insert!(𝒬, b)
+    haskey(ℬ, a) ? push!(ℬ[a], (a, b)) : (ℬ[a] = [(a, b)])
+    haskey(ℰ, b) ? push!(ℰ[b], (a, b)) : (ℰ[b] = [(a, b)])
+    lookup[(a, b)] = i
+  end
+
+  # sweep line algorithm
+  points = Vector{P}()
+  seginds = Vector{Vector{Int}}()
+  visited = Dict{P,Int}()
```

Comment:
I believe we can eliminate this `visited` auxiliary variable if we start the function with all points rounded. The first block of code in the function could round the coords of points and then orient the segments. Everything is done in terms of rounded coordinates.